legacyClaim/starknet/src/bridge.cairo

```diff
@@ -0,0 +1,165 @@
+
+    #[abi(embed_v0)]
+    impl OwnableMixinImpl = OwnableComponent::OwnableMixinImpl<ContractState>;
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+    impl UpgradeableInternalImpl = UpgradeableComponent::InternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        claimed: LegacyMap<EthAddress, bool>,
+        l1_bridge_address: EthAddress,
+        reward_token: ContractAddress,
+        reward_sponsor: ContractAddress,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        bridge_admin: ContractAddress,
+        l1_bridge_address: EthAddress,
+        reward_token: ContractAddress
+    ) {
+        self.ownable.initializer(bridge_admin);
+        self.l1_bridge_address.write(l1_bridge_address);
+        self.reward_token.write(reward_token);
+    }
+
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        ClaimRequestCompleted: ClaimRequestCompleted,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ClaimRequestCompleted {
+        #[key]
+        l1_address: EthAddress,
+        #[key]
+        l2_address: ContractAddress,
+        amount: u256
+    }
+
+    /// Process message from L1 to receive realm token.
+    ///
+    /// # Arguments
+    ///
+    /// `from_address` - L1 sender address, must be Realms L1 Bridge.
+    /// `req` - reward collection request.
+    ///
+    #[l1_handler]
+    fn withdraw_auto_from_l1(ref self: ContractState, from_address: felt252, req: Request) {
+        // ensure only the l1 bridge contract can cause this function to be called
+        assert(
+            self.l1_bridge_address.read().into() == from_address, 'Bridge: Caller not L1 Bridge'
+        );
+
+        // ensure the recipient's l1 and l2 addresses are non zero
+        assert(req.owner_l1.is_non_zero(), 'Bridge: zero l1 address');     
+        assert(req.owner_l2.is_non_zero(), 'Bridge: zero l2 address');     
+
+        // ensure the rightful l1 owner called this function
+        let (claim_l1_owner, claim_amount ) = claims_mapping(req.claim_id.into());
+        assert!(req.owner_l1.into() == claim_l1_owner, "Bridge: l1 caller not reward owner");
+
+        // ensure claim can't be done more than once
+        assert!(self.claimed.read(req.owner_l1) == false, "Bridge: l1 caller already claimed");        
+        self.claimed.write(req.owner_l1, true);
+
+        // transfer reward
+        assert!(
+            ERC20ABIDispatcher{contract_address: self.reward_token.read()}
+            .transfer_from(self.reward_sponsor.read(), req.owner_l2, claim_amount.into()),
```

Comment:
So the reward_sponsor will give unlimited allowance to the L2 bridge contract, to ensure this won't fail? Just asking, couldn't find this in a deploy script (not sure if it's part of the PR tbh).

Reply:
the approval can be limited to the total amount we expect to be claimed. havent added deploy script. will do so in a few

Reply:
Agreed, total amount expected is safest.